### ISSUE TITLE
Cannot assign a Thunderbolt port to a VM

### ISSUE THREAD

**Issue by DemiMarie** [QubesOS/qubes-issues#4353]
### Qubes OS version:

R4.0

### Affected component(s): 

dom0 kernel, qubes-core-admin

---

### Steps to reproduce the behavior:

Try to assign a Thunderbolt port to a VM.  This is incredibly useful for e.g. external graphics cards, which (to my understanding) are the only way to do gaming/GPGPU in Qubes without a security risk.

### Expected behavior:

Thunderbolt can be passed through in the usual way.

### Actual behavior:

No way to figure out which PCI device is the Thunderbolt device.

### General notes:

My understanding is that this makes doing GPGPU or gaming on Qubes in a supported and secure configuration impossible.

---

**Comment by andrewdavidwong** [comment]
Related issue: #2454

**Comment by marmarek** [comment]
Does it also happen if you boot the system with the device plugged in? If it doesn't work then either, the problem is somewhere else.
Anyway, I think just having CONFIG_PCI_HOTPLUG enabled isn't enough for hotplug to the VM work. Something needs to also (re-)configure IOMMU when you connect the device. AFAIK this isn't handled automatically right now.

As for "doesn’t attempt to do anything with them", Thunderbolt supports some kind of device connection policy. It should be possible to use that to prevent device to access any resources (especially DMA to system RAM) before it gets accepted (for example assigned to a VM. But I don't know details, including support for that in Linux and Xen. Without thing like this, enabling PCI hotplug is risky, because PCI devices generally (unless properly isolated using IOMMU) can access system RAM using DMA.

**Comment by DemiMarie** [comment]
I am not talking about assigning the individual Thunderbolt device, but
rather the entire controller.  Does that make sense?

On Sep 29, 2018 5:10 PM, "Marek Marczykowski-Górecki" <
notifications@github.com> wrote:

Does it also happen if you boot the system with the device plugged in? If
it doesn't work then either, the problem is somewhere else.
Anyway, I think just having CONFIG_PCI_HOTPLUG enabled isn't enough for
hotplug to the VM work. Something needs to also (re-)configure IOMMU when
you connect the device. AFAIK this isn't handled automatically right now.

As for "doesn’t attempt to do anything with them", Thunderbolt supports
some kind of device connection policy. It should be possible to use that to
prevent device to access any resources (especially DMA to system RAM)
before it gets accepted (for example assigned to a VM. But I don't know
details, including support for that in Linux and Xen. Without thing like
this, enabling PCI hotplug is risky, because PCI devices generally (unless
properly isolated using IOMMU) can access system RAM using DMA.

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub
<https://github.com/QubesOS/qubes-issues/issues/4353#issuecomment-425675791>,
or mute the thread

**Comment by marmarek** [comment]
Connected devices are separate PCI devices. Behind a bridge (separate for each port), but still a device, which needs to be properly guarded by IOMMU. Assigning just the bridge (which to be honest I'm not sure if Xen supports at all) isn't enough.

Anyway, you can try, `lspci` and `lspci -t` output in dom0 should be helpful in figuring out which device is connected where.

**Comment by DemiMarie** [comment]
This will be fixed when #6620 is.  Xen has safe support for PCIe hotplug as of [XSA-306](https://xenbits.xen.org/xsa/advisory-306.html).

**Comment by swenherweg** [comment]
Hello,
has anyone found a solution to enable Thunderbold 3?

**Comment by DemiMarie** [comment]
This one is up to @marmarek.  I am not sure how to test Xen’s IOMMU handling.

**Comment by dylangerdaly** [comment]
Was there ever any update to this? I bought a USB4 10GbE adapter and would like to test it in Qubes, guessing it's just flipping on `CONFIG_PCI_HOTPLUG`?

Plugging in an older TB3 NVMe Drive won't enumerate the PCI Device, UEFI Settings are all set to allow PCIe Tunneling

**Comment by marmarek** [comment]
No, it isn't just enabling hotplug in dom0 kernel, it's also ensuring IOMMU is correctly configured in Xen. And assigning it properly to sys-usb (or another). The last two parts are not implemented.

**Comment by dylangerdaly** [comment]
Wouldn't it be enabling hotplugging for sys-usb's kernel? I've passed the USB4 Controller to sys-usb already.

Also I thought Xen didn't care about IOMMU groupings? USB4 support in Qubes would be a great quality of life improvement.

How can I check Xen's IOMMU settings?

**Comment by DemiMarie** [comment]
Thunderbolt creates a PCIe tunnel.  That is a separate device.  Xen not caring about IOMMU groups is a vulnerability (which is not exploitable in most configurations I believe), not a feature.

**Comment by dylangerdaly** [comment]
Ahh okay, so it would just show up in dom0, yeah that's kind of scary.

**Comment by DemiMarie** [comment]
The PCIe tunnel is a separate PCIe device.  Not caring about IOMMU groups only allows insecure configurations to be created.  If a configuration works only because Xen is unaware of IOMMU groups, that configuration is insecure.

**Comment by marmarek** [comment]
Fortunately, with Thunderbolt, those devices usually get separate IOMMU group. But yes, that needs to be checked before allowing device to issue any DMA transactions. And most importantly, the new device that shows up needs to be moved out of dom0 (including appropriate IOMMU mapping) before enabling it.

**Comment by dylangerdaly** [comment]
I wonder if you booted with the Thunderbolt device plugged in, if UEFI would set it up for dom0? As I don’t really need Hot Plugging.

What steps would be required to get Thunderbolt to add PCIe devices to sys-usb?

How can I confirm what IOMMU groups are where etc? I’m guessing Xen blocks all DMA requests from all PCI Devices

**Comment by marmarek** [comment]
If you plug it before starting the system, you should have the PCI device present already (details might depend on specific firmware) and you can assign it to sys-usb. But the risk here is that your firmware might want to interact with the device and while some do that safely (including IOMMU setup), not all of them do. In that case, if the device is malicious, it could compromise the firmware and/or Xen/dom0 during boot. You can read more about some of those attacks at https://thunderspy.io/
And also, if you unplug the device while the system is running, it will get confused (in my system it doesn't fully crash, just throw a lot of errors, but again - this depends on specific hardware).

**Comment by dylangerdaly** [comment]
I’ll give this a shot when I get back to my desk, I’m really curious to see if I can get anything close to 8Gbps to sys-net, my plan is to add the TB3 10Gbps NIC to sys-net directly and run tests, attach a VM downstream and test again to see how much the bottleneck is Qubes

Outside of that backing Qubes up with a TB3 NVMe enclosure sounds great.

Hopefully booting with the devices plugged in works

**Comment by DemiMarie** [comment]
> If you plug it before starting the system, you should have the PCI device present already (details might depend on specific firmware)

I thought setting up Thunderbolt tunnels usually required user interaction, at least if the firmware is configured correctly.  I know that some desktop environments will prompt the user to set up the PCI tunnel, with the default being not to.

**Comment by marmarek** [comment]
As I said, depends on the firmware.
